### PR TITLE
Add auth script execution supervisor

### PR DESCRIPTION
**Pull request by janakamarasena**
This PR introduces an adaptive auth script execution supervisor that is capable of limiting the execution time of the script.

If required the following can be configured (in the deployment.toml);
Number of threads for the supervisor:
Default:1
```
[authentication.adaptive.execution_supervisor]
thread_count = "1"
```

Allowed script execution time in milliseconds:
Default:500ms
```
[authentication.adaptive.execution_supervisor]
timeout = "500ms"
```

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        currentScriptExecutions.put(identifier, monitoringFuture);
+    }
+
+    /**
+     * Mark adaptive auth script execution as complete and stop monitoring.
+     *
+     * @param identifier Monitoring task identifier.
+     * @return Total elapsed time of adaptive auth execution in the current thread.
+     */
+    public long competed(String identifier) {
```
Comment (omindu) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java:95]: `competed` typo ?

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/org.wso2.carbon.identity.core.server.feature.default.json [hunk 2]
```diff
@@ -577,6 +577,8 @@
   "authentication.adaptive.long_wait.page_refresh_interval": "500ms",
   "authentication.adaptive.long_wait.timout": "10s",
   "authentication.adaptive.long_wait.prompt": false,
+  "authentication.adaptive.execution_supervisor.thread_count": 1,
+  "authentication.adaptive.execution_supervisor.timeout": 500,
```
Comment (senthalan) [inline features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/org.wso2.carbon.identity.core.server.feature.default.json:581]: let's have an entry in `org.wso2.carbon.identity.core.server.feature.unit-resolve.json` to specify that this config is in `ms`.

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        if (taskExecutionRateInMillis > timeoutInMillis) {
+            taskExecutionRateInMillis = timeoutInMillis;
+        }
+
+        init();
+    }
+
+    private void init() {
+
+        monitoringService = new ScheduledThreadPoolExecutor(THREAD_COUNT, r -> new Thread(r, JS_EXECUTION_MONITOR));
```
Comment (madurangasiriwardena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java:57]: do we really need a separate init method?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java (hunk 1 above)
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java]: fixed

File: features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/org.wso2.carbon.identity.core.server.feature.default.json (hunk 2 above)
Comment (janakamarasena) [reply features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/org.wso2.carbon.identity.core.server.feature.default.json:581]: added

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java (hunk 3 above)
Comment (janakamarasena) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java:52]: updated

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        @Override
+        public void run() {
+
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(String.format("JS execution monitoring task running. Thread: %s, service " +
+                "provider: %s, tenant: %s.", originalThread.getName(), serviceProvider, tenantDomain));
+            }
+
+            long elapsedTime  = getTotalElapsedTime();
+            if(elapsedTime > TIMEOUT_IN_MILLIS) {
```
Comment (janakamarasena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java:135]: Formatting

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java [hunk 5]
```diff
@@ -983,8 +1015,13 @@ public Object evaluate(AuthenticationContext authenticationContext, Function<JSO
 
                     CompiledScript compiledScript = compilable.compile(jsFunction.getSource());
                     JSObject builderFunction = (JSObject) compiledScript.eval();
-                    result = jsConsumer.apply(builderFunction);
 
+                    long elapsedTime = getAuthScriptExecutionElapsedTime(authenticationContext);
+                    getJSExecutionSupervisor().monitor(identifier, authenticationContext.getServiceProviderName(),
+                            authenticationContext.getTenantDomain(), elapsedTime);
+                    result = jsConsumer.apply(builderFunction);
+                    elapsedTime = getJSExecutionSupervisor().completed(identifier);
```
Comment (janakamarasena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JsGraphBuilder.java:1023]: Add try finally

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+     * @param tenantDomain Tenant domain.
+     * @param elapsedTimeInMillis Elapsed time in milliseconds if a previous
+     *                            adaptive auth execution happened in the same thread.
+     */
+    public void monitor(String identifier, String serviceProvider, String tenantDomain, long elapsedTimeInMillis) {
+
+        MonitoringTask monitoringTask = new MonitoringTask(Thread.currentThread(), identifier, serviceProvider,
+                tenantDomain, elapsedTimeInMillis);
+        monitoredTasks.put(identifier, monitoringTask);
+        ScheduledFuture<?> monitoringFuture = monitoringService.
```
Comment (janakamarasena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java:77]: Add to a single map

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+     */
+    public long completed(String identifier) {
+
+        MonitoringTask task = monitoredTasks.remove(identifier);
+        ScheduledFuture<?> monitoredFuture = currentScriptExecutions.remove(identifier);
+        if (monitoredFuture == null) {
+            // Nothing to be done as there was no such task with the given identifier.
+            return 0L;
+        }
+        monitoredFuture.cancel(false);
```
Comment (janakamarasena) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/config/model/graph/JSExecutionSupervisor.java:98]: Set to true

**Review by madurangasiriwardena — APPROVED** [review]

**Pull request merged**